### PR TITLE
feature: add rust-fnv@v1.0.6

### PR DESCRIPTION
**Pull request by sammyne**
### Description

Port in rust-fnv@1.0.6

### Crate Meta

- **rust-fnv**:
- **1.0.6**:
- **MIT/Apache-2.0**:
- **2ca7554508beb26d7b3f817008d0e58da367d227**:
- **v1.1.1**:

### Diff of Changes

```diff
diff -Naur '--exclude=.git' '--exclude=sgx-tests' crates/rust-fnv/Cargo.toml ../rust-fnv/Cargo.toml
--- crates/rust-fnv/Cargo.toml	2020-04-03 11:38:11.184539990 +0000
+++ ../rust-fnv/Cargo.toml	2020-04-03 11:44:22.459337505 +0000
@@ -7,17 +7,7 @@
 readme = "README.md"
 repository = "https://github.com/servo/rust-fnv"
 documentation = "https://doc.servo.org/fnv/"
-edition = "2018"
 
 [lib]
 name = "fnv"
 path = "lib.rs"
-
-[features]
-default = ["mesalock_sgx"]
-mesalock_sgx = ["sgx_tstd"]
-enclave_unit_test = ["sgx_tunittest"]
-
-[dependencies]
-sgx_tstd        = { git = "https://github.com/apache/incubator-teaclave-sgx-sdk.git", tag = "v1.1.1", optional = true }
-sgx_tunittest   = { git = "https://github.com/apache/incubator-teaclave-sgx-sdk.git", tag = "v1.1.1", optional = true }
diff -Naur '--exclude=.git' '--exclude=sgx-tests' crates/rust-fnv/lib.rs ../rust-fnv/lib.rs
--- crates/rust-fnv/lib.rs	2020-04-03 11:38:11.185540059 +0000
+++ ../rust-fnv/lib.rs	2020-04-03 11:44:22.460337576 +0000
@@ -64,11 +64,6 @@
 //! [faq]: https://www.rust-lang.org/en-US/faq.html#why-are-rusts-hashmaps-slow
 //! [graphs]: http://cglab.ca/~abeinges/blah/hash-rs/
 
-#![cfg_attr(all(feature = "mesalock_sgx",
-                not(target_env = "sgx")), no_std)]
-#![cfg_attr(all(target_env = "sgx", target_vendor = "mesalock"), feature(rustc_private))]
-#[cfg(all(feature = "mesalock_sgx", not(target_env = "sgx")))]
-extern crate sgx_tstd as std;
 
 use std::default::Default;
 use std::hash::{Hasher, BuildHasherDefault};
@@ -126,19 +121,11 @@
 pub type FnvHashSet<T> = HashSet<T, FnvBuildHasher>;
 
 
-#[cfg(feature = "enclave_unit_test")]
-pub mod test {
-    use std::prelude::v1::*;
-
+#[cfg(test)]
+mod test {
     use super::*;
     use std::hash::Hasher;
 
-    use sgx_tunittest::*;
-
-    pub fn run_tests() {
-        rsgx_unit_tests!(basic_tests);
-    }
-
     fn fnv1a(bytes: &[u8]) -> u64 {
         let mut hasher = FnvHasher::default();
         hasher.write(bytes);
@@ -153,7 +140,7 @@
         (0..500).flat_map(|_| bytes.iter().cloned()).collect()
     }
 
-    //#[test]
+    #[test]
     fn basic_tests() {
         assert_eq!(fnv1a(b""), 0xcbf29ce484222325);
         assert_eq!(fnv1a(b"a"), 0xaf63dc4c8601ec8c);
```

### CODE REVIEW

**Comment by mssun** [comment]
Can you use the test template for the unit/functional tests? https://github.com/universal-secure-computing-community/crates-sgx/tree/sgx1.1.1/utils/sgx-tests-template

The new template removes some redundant header files, irrelevant code and provide a minimal framework.

**Comment by sammyne** [comment]
Check it again~ I have removed the redundant files now 😄  @mssun

File: crates/rust-fnv/.drone.yml [hunk 1]
```diff
@@ -0,0 +1,237 @@
+kind: pipeline
```
Comment (mssun) [inline crates/rust-fnv/.drone.yml]: Delete this file.

File: crates/rust-fnv/sgx-tests/enclave/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#![cfg_attr(not(target_env = "sgx"), no_std)]
+#![cfg_attr(target_env = "sgx", feature(rustc_private))]
+
+#[cfg(not(target_env = "sgx"))]
+extern crate sgx_tstd as std;
+
+#[no_mangle]
+pub extern "C" fn ecall_run_tests() {
+    fnv::test::run_tests();
+}
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/enclave/src/lib.rs]: Add a newline here.

File: crates/rust-fnv/sgx-tests/enclave/Makefile [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Copyright (C) 2017-2018 Baidu, Inc. All Rights Reserved.
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/enclave/Makefile]: Copyright header is not needed. See the template. We want to make things consistent.

File: crates/rust-fnv/diff.txt [hunk 4]
```diff
@@ -0,0 +1,7425 @@
+diff -Naur '--exclude=.git' crates/rust-fnv/.drone.yml ../rust-fnv/.drone.yml
```
Comment (mssun) [inline crates/rust-fnv/diff.txt]: No need to put a diff .txt here. Copy it to the PR to make reviewing/auditing easier.

Also, we only need diff of changes on the original project. Please exclude additional  the sgx-tests directory.

**Comment by mssun** [comment]
Is you change based on https://github.com/servo/rust-fnv/tree/v1.0.6 (948b9d3b3e71743a2884165af6d18bfdc9095f76)? I found a lot differences.

**Comment by sammyne** [comment]
> Is you change based on https://github.com/servo/rust-fnv/tree/v1.0.6 (948b9d3b3e71743a2884165af6d18bfdc9095f76)? I found a lot differences.

I have rebased on https://github.com/servo/rust-fnv/tree/v1.0.6 now~

File: crates/rust-fnv/sgx-tests/app/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "app"
+version = "1.0.0"
+authors = ["Baidu"]
+build = "build.rs"
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/app/Cargo.toml:4]: Remove `authors`, add edition. Same with this: https://github.com/universal-secure-computing-community/crates-sgx/blob/sgx1.1.1/crates/itoa/sgx-tests/app/Cargo.toml

File: crates/rust-fnv/sgx-tests/enclave/Enclave.edl [hunk 6]
```diff
@@ -0,0 +1,43 @@
+/*
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/enclave/Enclave.edl]: Remove this header.

**Comment by sammyne** [comment]
Now, another update as recommended~  @mssun

File: crates/rust-fnv/sgx-tests/Makefile [hunk 7]
```diff
@@ -0,0 +1,174 @@
+# Copyright (C) 2017-2018 Baidu, Inc. All Rights Reserved.
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/Makefile]: Please also delete this header.

File: crates/rust-fnv/sgx-tests/app/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,86 @@
+// Copyright (C) 2017-2018 Baidu, Inc. All Rights Reserved.
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/app/src/main.rs]: Please also delete this header.

File: crates/rust-fnv/sgx-tests/app/build.rs [hunk 9]
```diff
@@ -0,0 +1,47 @@
+// Copyright (C) 2017-2018 Baidu, Inc. All Rights Reserved.
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/app/build.rs]: Please also delete this header.

File: crates/rust-fnv/sgx-tests/enclave/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "run-tests-enclave"
+version = "1.0.0"
+authors = ["Baidu"]
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/enclave/Cargo.toml]: Please check whether this `authors` is same with the `Cargo.toml` in the template.

File: crates/rust-fnv/sgx-tests/enclave/Makefile [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+Rust_Target_Path := $(CURDIR)
+
+.PHONY: all
+
+all: $(Rust_Enclave_Name)
+
+$(Rust_Enclave_Name): $(Rust_Enclave_Files)
+ifeq ($(XARGO_SGX), 1)
+	RUST_TARGET_PATH=$(Rust_Target_Path) xargo build --target x86_64-unknown-linux-sgx --release
+	cp ./target/x86_64-unknown-linux-sgx/release/libhelloworldsampleenclave.a ../lib/libenclave.a
```
Comment (mssun) [inline crates/rust-fnv/sgx-tests/enclave/Makefile]: What's `helloworldsampleenclave`? I guess I have changed this line.

**Comment by mssun** [comment]
Can you verify that the utility files of `sgx-tests` are same with template? This is important because we want to make sure all crates are created in the same process and have similar structure.

**Comment by sammyne** [comment]
> Can you verify that the utility files of `sgx-tests` are same with template? This is important because we want to make sure all crates are created in the same process and have similar structure.

Now, it should be consistent = = @mssun

**Comment by mssun** [comment]
Merged. Thanks!

**Pull request merged**